api/src/main/java/org/openmrs/module/fhirExtension/validators/DiagnosticReportBasedOnValidator.java

```diff
@@ -0,0 +1,46 @@
+package org.openmrs.module.fhirExtension.validators;
+
+import ca.uhn.fhir.rest.server.exceptions.UnprocessableEntityException;
+import org.hl7.fhir.r4.model.DiagnosticReport;
+import org.hl7.fhir.r4.model.OperationOutcome;
+import org.openmrs.Order;
+import org.openmrs.Patient;
+import org.openmrs.api.OrderService;
+import org.openmrs.module.fhir2.model.FhirDiagnosticReport;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+
+@Component
+public class DiagnosticReportBasedOnValidator {
+	
+	static final String INVALID_ORDER_ERROR_MESSAGE = "Given lab order is not prescribed by the doctor";
+	
+	@Autowired
+	private OrderService orderService;
+	
+	public void validate(FhirDiagnosticReport fhirDiagnosticReport, DiagnosticReport diagnosticReport) {
+		if (diagnosticReport.getBasedOn() != null && diagnosticReport.getBasedOn().get(0).getReference() != null)
+			validateBasedOn(fhirDiagnosticReport);
+	}
+	
+	private void validateBasedOn(FhirDiagnosticReport fhirDiagnosticReport) {
+        Patient patient = fhirDiagnosticReport.getSubject();
+        Integer conceptId = fhirDiagnosticReport.getCode().getId();
+        List<Order> allOrders = orderService.getAllOrdersByPatient(patient);
+        long matchingOrdersCount = allOrders.stream().filter((order) -> order.isActive()).filter((order) -> order.getConcept().getId() == conceptId).count();
```

Comment:
lambda and null safe equals

`long matchingOrdersCount = allOrders.stream().filter(Order::isActive).filter((order) -> order.getConcept().getId().equals(conceptId)).count();`